### ISSUE TITLE
Persistent high CPU usage after upgrading to v4.x

### ISSUE THREAD

**Issue by bariscicek** [zcash/zcash#4858]
### Describe the issue
After upgrading to v4.0.0 and v4.1.0, zcash node started to have persistent high CPU usage. Node initializes fine, but after loading wallet CPU usage goes top.

#### If so, please list the steps to reproduce below:
1. Start zcash node.
2. Wait for a while and see CPU usage goes 100%
 

### Expected behaviour
CPU usage profile should be similar to v3.1

### The version of Zcash you were using:
v4.1.0

### Any extra information that might be useful in the debugging process.
Console:
```
     Downloading blocks | 1035930 (1036250 headers) / ~1037570 (99%)
     Downloading blocks | 1035930 (1036250 headers) / ~1038180 (99%)
                        | [                                                  ]
           Next upgrade | Canopy at block height 1046400, in around 9 days
            Connections | 0
  Network solution rate | 7.109 GSol/s
```

Tail of debug.log:
```
Nov 11 04:18:33.558  INFO main: keypool reserve 305394
Nov 11 04:18:33.560  INFO main: keypool keep 305394
Nov 11 04:59:12.845  INFO main: keypool reserve 305395
Nov 11 04:59:12.847  INFO main: keypool keep 305395
Nov 11 05:03:46.695  INFO main: keypool reserve 305396
Nov 11 05:03:46.697  INFO main: keypool keep 305396
Nov 11 05:17:23.462  INFO main: connect() to [2001:a18:0:b23::6]:8233 failed: Network is unreachable (101)
Nov 11 05:30:55.828  INFO main: keypool reserve 305397
Nov 11 05:30:55.830  INFO main: keypool keep 305397
Nov 11 06:25:19.009  INFO main: keypool reserve 305398
Nov 11 06:25:19.011  INFO main: keypool keep 305398
Nov 11 06:57:12.652  INFO main: keypool reserve 305399
Nov 11 06:57:12.656  INFO main: keypool keep 305399
Nov 11 07:06:56.305  INFO main: connect() to [2a02:8109:88c0:a6a::472a]:8233 failed: Network is unreachable (101)
Nov 11 07:07:53.484  INFO main: connect() to [2a03:1b20:1:e011::a19e]:8233 failed: Network is unreachable (101)
Nov 11 07:08:11.306  INFO main: connect() to [2600:1700:e860:8200:50b4:330e:7917:d1f9]:8233 failed: Network is unreachable (101)
Nov 11 07:10:02.723  INFO main: connect() to 178.175.148.25:8233 failed after select(): No route to host (113)
Nov 11 07:10:49.603  INFO main: connect() to 124.65.151.98:8233 failed after select(): Connection refused (111)
Nov 11 07:12:03.625  INFO main: connect() to 23.129.64.234:8233 failed after select(): Connection refused (111)
Nov 11 07:12:04.274  INFO main: connect() to 79.110.53.130:8233 failed after select(): Connection refused (111)
Nov 11 07:12:11.384  INFO main: connect() to [2400:2410:9821:3700:75b2:3475:70d1:9741]:8233 failed: Network is unreachable (101)
Nov 11 07:29:09.276  INFO main: keypool reserve 305400
Nov 11 07:29:09.280  INFO main: keypool keep 305400
Nov 11 07:33:43.766  INFO main: keypool reserve 305401
Nov 11 07:33:43.770  INFO main: keypool keep 305401
Nov 11 08:55:42.858  INFO main: keypool reserve 305402
Nov 11 08:55:42.860  INFO main: keypool keep 305402
Nov 11 09:45:38.878  INFO main: keypool reserve 305403
Nov 11 09:45:38.880  INFO main: keypool keep 305403
```

**Comment by steven-ecc** [comment]
What's operating system are you running zcashd on? Did you compile from source or install our binary?

**Comment by bariscicek** [comment]
@steven-ecc tried both compiled and ubuntu binary. both has same issue.

**Comment by bariscicek** [comment]
@mdr0id it is keeping CPU usage high regardless how long you wait and connections drop to 0.

**Comment by mdr0id** [comment]
Do you have any special configs in your zcash.conf?

**Comment by bariscicek** [comment]
@mdr0id no, just `addnode=mainnet.z.cash`.

**Comment by defuse** [comment]
While the node is running and using 100% CPU, can you run `netstat -atn`? The output will help us rule out a denial of service attack coming from a peer.

**Comment by mdr0id** [comment]
@bariscicek Also, is this running on the native host or inside a containerized env?

**Comment by bariscicek** [comment]
@mdr0id running on native host.

@defuse I will provide the info, please note that this is only happening on v4.x not on v3.x.

**Comment by bariscicek** [comment]
on v3.x there is significant CPU usage, but I do not see connections dropping 0. I am thinking on v4.x trying to find peers put more overhead on top of validation. Is there any way to see why I have 0 connections on v4.1, but having 8 on v3.1 on same network configuration only client version changes?

**Comment by mdr0id** [comment]
@bariscicek #4864 Is now in master and should address the mentioned issue. Please let us know if you have issues with initial node syncs and peering.

**Comment by daira** [comment]
I actually don't know whether it will address this specific issue, but there's a very good chance that it will.

**Comment by bariscicek** [comment]
Now connection is stable at 8, though it is slow to catch height. Attached is the debug.log. Since Canopy block is almost there, urgent help is appreciated. By the way machine specs as follows:
processor	: 35
vendor_id	: GenuineIntel
cpu family	: 6
model		: 85
model name	: Intel(R) Xeon(R) Platinum 8124M CPU @ 3.00GHz
stepping	: 3
microcode	: 0x1000141
cpu MHz		: 3388.256
cache size	: 25344 KB

It's a 36 core virtual machine.
[debug.log](https://github.com/zcash/zcash/files/5550147/debug.log)

**Comment by mdr0id** [comment]
@bariscicek How much RAM is on this system and what is the operating system?

**Comment by daira** [comment]
That `debug.log` appears to show a synced node. What exactly is slow, and can you quantify how slow?

**Comment by bariscicek** [comment]
@daira after these logs it stalls for say, 30 mins so becomes out of sync.

**Comment by josef-v** [comment]
We have also observed a higher CPU usage after upgrading from v4.0.0 to v4.1.0 (the only change in the environment). 
We are running our nodes in docker based on debian 10. 
Binaries are custom built using the `zcutil/build.sh` script with `--disable-tests` configure flag.
Switch from v4.0.0 to v4.1.0 can be seen here (around 14:30, but it's quite evident):
![zcashcpu](https://user-images.githubusercontent.com/52073977/99654759-1523f280-2a5b-11eb-806d-97f55a256273.png)

We have also encountered a problem we have never encountered before and that is `http work queue depth exceeded`. This has never happened before and we made no changes to our code so I am suspicious it might be related to the CPU usage.
